### ISSUE TITLE
Don't try unsupported operations

### ISSUE THREAD

**Issue by daurnimator** [wahern/cqueues#21]
e.g. When using UDP, TCP options are attempted:

```
$ strace -e 'getsockopt' lua <<< 'cs=require"cqueues.socket"; cs.connect{host="localhost", port=2323,type=cs.SOCK_DGRAM}'
getsockopt(4, SOL_SOCKET, SO_REUSEADDR, [1], [4]) = 0
getsockopt(4, SOL_SOCKET, SO_REUSEPORT, [0], [4]) = 0
getsockopt(4, SOL_TCP, TCP_NODELAY, 0x7fff86f69ac8, 0x7fff86f69acc) = -1 EOPNOTSUPP (Operation not supported)
getsockopt(4, SOL_TCP, TCP_CORK, 0x7fff86f69ac8, 0x7fff86f69acc) = -1 EOPNOTSUPP (Operation not supported)
getsockopt(4, SOL_IPV6, IPV6_V6ONLY, 0x7fff86f69ac8, 0x7fff86f69acc) = -1 EOPNOTSUPP (Operation not supported)
+++ exited with 0 +++
```

Don't try and use SOL_TCP options when socket is not `AF_INET`, `SOCK_STREAM`, `0`

**Comment by wahern** [comment]
Can you try this patch?

```
diff --git a/src/lib/socket.c b/src/lib/socket.c
index c0fad30..4bad5b1 100644
--- a/src/lib/socket.c
+++ b/src/lib/socket.c
@@ -1212,7 +1212,7 @@ error:
 
 static int so_socket_(struct socket *so) {
 	struct stat st;
-	int error;
+	int mask, error;
 
 	if (!so->host)
 		return SO_ENOHOST;
@@ -1222,7 +1222,8 @@ static int so_socket_(struct socket *so) {
 	if (-1 == (so->fd = so_socket(so->host->ai_family, so->host->ai_socktype, &so->opts, &error)))
 		return error;
 
-	so->flags = so_getfl(so->fd, ~0);
+	mask = so_type2mask(so->host->ai_family, so->host->ai_socktype, 0);
+	so->flags = so_getfl(so->fd, mask);
 
 	if (0 != fstat(so->fd, &st))
 		return errno;
```

**Comment by daurnimator** [comment]
That works :) nice fix!

Related issue: `fdopen` on non-socket descriptors fails

```lua
assert(cs.fdopen(0))
```
Fails with `88`. i.e. ENOTSOCK: "Operation not supported"


This seems to be inside the `so_rstfl` call.

`so_type2mask` might need to additionally take a `mode_t`?

**Comment by daurnimator** [comment]
This also causes issues with `:recvfd`:

```lua
local ct = require "cqueues.thread"
local t, p = ct.start(function(p)
        print("GOT", p:recvfd())
end)
p:sendfd("hi", io.stdin)
print(t:join())
```

Which shows:
```
true    ../../threads-cq-serpent.lua:8: socket:recvfd: Operation not supported
```

**Comment by daurnimator** [comment]
This should be easier to fix now that the socket object has `mode`, `domain` and `type` fields.

**Comment by daurnimator** [comment]
patch to add to this issue:
```diff
diff --git a/src/lib/socket.c b/src/lib/socket.c
index f39d65b..29e9f76 100644
--- a/src/lib/socket.c
+++ b/src/lib/socket.c
@@ -2788,6 +2788,9 @@ int so_peerfd(struct socket *so) {
 
 
 int so_uncork(struct socket *so) {
+       if (so->protocol != IPPROTO_TCP) {
+               return EOPNOTSUPP;
+       }
        return so_nopush(so->fd, 0);
 } /* so_uncork() */
 
```

**Comment by wahern** [comment]
Rather than adding a bunch of logic for each option that records what kinds of sockets it's compatible with, I've been leaning toward refactoring the way options are specified.

Right now they're either enabled or disabled. That was fine when the set of options was only O_NONBLOCK, O_CLOEXEC, disabling SIGPIPE, etc. It made sense to make sure that they were always set appropriately, especially when instantiating a socket from an existing descriptor. And options like SO_REUSEADDR were always broadly applicable, so I carried forward the boolean enable/disable behavior.

But most new options should behave the way that the ```v6only``` option is handled, where the default is to ignore it entirely. So I want to refactor ```reuseport```, ```nopush```, ```oobinline```, etc.

**Comment by daurnimator** [comment]
I consider this issue fixed now with 9be415a
